include/envoy/common/conn_pool.h

```diff
@@ -64,6 +64,22 @@ class Instance {
    */
   virtual void drainConnections() PURE;
 
+  /**
+   * Called when a connection pool has no pending requests or busy connections for a configured
+   * timeout duration.
+   */
+  using IdlePoolTimeoutCb = std::function<void()>;
+
+  /**
+   * Register a callback that gets called when the connection pool has had no pending requests or
+   * busy connections for a configured timeout duration.
+   *
+   * NOTE: The "idle" state is not the same as the "drained" state. A pool may be drained before it
+   * has a chance to go idle (e.g. a host is removed), and a pool is drained after it has
+   * gone idle.
+   */
+  virtual void addIdlePoolTimeoutCallback(IdlePoolTimeoutCb cb) PURE;
```

Comment:
I know we discussed this before, but it's been a while so I don't remember where we landed with this. Conceptually, it feels broken that we have "drain callbacks" as well as "idle callbacks."

Conceptually these seem like the same thing. How is idle different from draining?

Could we model this differently by doing something like the following:
1) Having draining just be a bool state (yes/no).
2) Have idle callbacks and timeout be independently setable.
3) Draining is basically just setting the drain boolean and setting the idle timeout to 0?

cc @alyssawilk who I would also like to look at this since she has been in this code a lot recently.

Reply:
Just getting back to this...

In this PR, the idle timer starts whenever there are no pending or active clients for the pool. When the timer fires, the pool is drained.

So the sequence is currently:
1. No active or pending clients on the pool, pool idle timer is set
1. Idle timer fires, idle callbacks called
1. Cluster manager drains the pool
1. All connections on pool are closed, drain callbacks called
1. Cluster manager deletes pool

I think in order to make your suggestion work, the definition of the start of the "idle" interval would have to be after all connections have hit their connection idle timeouts. Since "drained" requires that all connections are closed, in order for there to be a single callback, the idle timeout will also have to be fired at that time or later.

This has the downside that a user will have to appropriately configure both the connection idle timeout and the pool idle timeout on a connection pool. Leaving the default connection idle timeout for HTTP would mean that the timeout would be 1 hour + the configured pool idle timeout. If we're okay with asking the user to configure both timeouts, I can make the appropriate changes to this PR and merge the drained callbacks and the idle callbacks.

The new sequence for an idle timeout would be:
1. All connections on the pool hit their idle timeout, pool idle timer is set
1. Pool idle timer fires, idle callbacks called
1. Cluster manager deletes the pool

Reply:
So I feel strongly that we should merge drain/idle. IMO it's the same thing. But I'm still confused. Isn't "draining" just setting an idle timeout of 0 as it is implemented today? Why can't the explicit drain process just reset idle timeout to 0 regardless of what it was configured for?

Reply:
The current implementation explicitly closes all connections and waits for the connection count to reach zero before calling the drain callbacks, so it's not exactly setting the connection idle timeout to 0.

In this PR I introduced a pool idle timeout independent of the individual connection timeouts, but I think what you're suggesting is that the connection idle timeouts should be good enough.

If we were to rely on the connection idle timeouts only, I'm not sure what to do with the case where all of the connections in a pool close for reasons other than an idle timeout (e.g. a momentary loss of connectivity with the upstream). If we were to rely solely on the number of active connections reaching zero to determine when the pool is "idle", then we might be prematurely deleting and recreating pools if the connections in the pool error out. Maybe it's fine to delete and recreate pools in that case?

Reply:
I can go back and look at the code again, but I'm nearly positive that the definition of drained has nothing to do with connection idle timeouts and only whether there are active/pending requests. This was implementing a long, long time ago before there even were connection idle timeouts. Are you sure your analysis is correct on what drain does?

Reply:
You're right that it has nothing to do with connection idle timeouts -- I thought you were referring to the connection idle timeouts when you said "Isn't 'draining' just setting an idle timeout of 0 as it is implemented today?"

My understanding of the drain process is as follows (in `ConnPoolImplBase::drainConnectionsImpl`):

1. Close all connections without active requests (`closeIdleConnectionsForDrainingPool`)
1. Move all busy/ready clients to the DRAINING state
1. After certain events (e.g. request completion) close all connections without active requests
1. When all connections have been closed, call the drained callbacks

My understanding of the code is that a pool isn't drained until it has no pending or active requests *and* its connections have been closed.

Reply:
Right but the point is connections are actively closed when they have no active/pending requests. This means we don't wait for connection idle. Given this, I'm pretty sure it's effectively just setting idle timeout to 0?

Reply:
Yes, I agree that it's effectively setting the connection idle timeout to 0.

So if I'm understanding correctly:
* the idle callbacks should trigger based on the connection idle timeout of the last connection in a pool
* the idle and drained callbacks should both be added via the same API
* starting to drain should be setting a draining boolean

The question I have now is: what should we do about a pool whose connections are closed prior to the connection idle timeout -- for example, the remote endpoint closes all connections? Is it acceptable to tear down the pool at that point, even if that means we might recreate the pool immediately afterwards?

Reply:
> Is it acceptable to tear down the pool at that point, even if that means we might recreate the pool immediately afterwards?

I don't see a problem with this, but if you want you could guard this functionality with whether the draining boolean is set? Up to you.